llarp/rpc/lokid_rpc_client.cpp

```diff
@@ -0,0 +1,202 @@
+              LogWarn("failed to update service node list");
+              return;
+            }
+            if (data.size() < 2)
+            {
+              LogWarn("lokid gave empty reply for service node list");
+              return;
+            }
+            try
+            {
+              self->HandleGotServiceNodeList(std::move(data[1]));
+            }
+            catch (std::exception& ex)
+            {
+              LogError("failed to process service node list: ", ex.what());
+            }
+          },
+          request.dump());
+    }
+
+    void
+    LokidRpcClient::Connected()
+    {
+      constexpr auto PingInterval = 1min;
+      constexpr auto NodeListUpdateInterval = 30s;
+
+      LogInfo("we connected to lokid [", *m_Connection, "]");
+      Command("admin.lokinet_ping");
+      m_lokiMQ->add_timer(
+          [self = shared_from_this()]() { self->Command("admin.lokinet_ping"); }, PingInterval);
+      m_lokiMQ->add_timer(
+          [self = shared_from_this()]() { self->UpdateServiceNodeList(); }, NodeListUpdateInterval);
+      UpdateServiceNodeList();
+    }
+
+    void
+    LokidRpcClient::HandleGotServiceNodeList(std::string data)
+    {
+      auto j = nlohmann::json::parse(std::move(data));
+      {
+        const auto itr = j.find("block_hash");
+        if (itr != j.end())
+        {
+          m_CurrentBlockHash = itr->get<std::string>();
+        }
+      }
+      {
+        const auto itr = j.find("unchanged");
+        if (itr != j.end())
+        {
+          if (itr->get<bool>())
+          {
+            LogDebug("service node list unchanged");
+            return;
+          }
+        }
+      }
+
+      std::vector<RouterID> nodeList;
+      {
+        const auto itr = j.find("service_node_states");
+        if (itr != j.end() and itr->is_array())
+        {
+          for (auto j_itr = itr->begin(); j_itr != itr->end(); j_itr++)
+          {
+            const auto ed_itr = j_itr->find("pubkey_ed25519");
+            if (ed_itr == j_itr->end() or not ed_itr->is_string())
+              continue;
+            RouterID rid;
+            if (rid.FromHex(ed_itr->get<std::string>()))
+              nodeList.emplace_back(std::move(rid));
+          }
+        }
+      }
+
+      if (nodeList.empty())
+      {
+        LogWarn("got empty service node list from lokid");
+        return;
+      }
+      // inform router about the new list
+      LogicCall(m_Router->logic(), [r = m_Router, nodeList]() { r->SetRouterWhitelist(nodeList); });
+    }
+
+    std::optional<SecretKey>
+    LokidRpcClient::ObtainIdentityKey()
+    {
+      std::promise<std::optional<SecretKey>> promise;
+
+      Request(
+          "admin.get_service_privkeys",
+          [self = shared_from_this(), &promise](bool success, std::vector<std::string> data) {
+            if (not success)
+            {
+              LogError("failed to get private key");
+              promise.set_value(std::nullopt);
+              return;
```

Comment:
Perhaps we can change this to throw on error; right now we fail to set a value, and that trickles back with empty/false returns to `Router::Configure` which then throws.  If we throw via the promise from here we can accomplish the same thing, but provide some more context.

This ought to work:
```C++
      std::promise<SecretKey> promise;

      Request(
          "admin.get_service_privkeys",
          [self = shared_from_this(), &promise](bool success, std::vector<std::string> data) {
            try {
              // all the current code, except the `LogError/promise.set_value(std::nullopt)/return`
              // bits replaced replaced with `throw std::runtime_error{"failed to get private key because [Jesus]"}`;
              // ...
            } catch (...) {
              promise.set_exception(std::current_exception());
            }
          }
```